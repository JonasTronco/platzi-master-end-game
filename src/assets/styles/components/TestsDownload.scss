@import '../main.scss';

.tests-download {
  @include colStaSta;

  &__options {
    @include flexEndCen;
    width: 100%;
    margin-bottom: $space_x3;
  }

  .btn,
  .btn--disabled {
    align-self: flex-end;
    margin-left: $space_x6;
  }

  .message {
    .btn,
    .btn--disabled {
      align-self: center;
      margin-left: 0;
    }
  }
}
