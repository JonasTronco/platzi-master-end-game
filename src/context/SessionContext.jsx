import { useState, useEffect, createContext, useMemo, useContext } from 'react';
import Router from 'next/router';

const SessionContext = createContext();
export function SessionProvider(props) {
  const [session, setSession] = useState(null);

  useEffect(() => {
    try {
      setSession(JSON.parse(localStorage.getItem('session')));
    } catch (error) {
      Router.push('/');
    }
  }, []);

  const value = useMemo(
    () => ({
      session,
    }),
    [session],
  );

  return <SessionContext.Provider value={value} {...props} />;
}

export function useSession() {
  const context = useContext(SessionContext);
  if (!context) {
    throw new Error('No Session Context Provider');
  }
  return context;
}
