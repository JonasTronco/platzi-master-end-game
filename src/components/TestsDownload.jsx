import { useState } from 'react';
import axios from 'axios';
import { useSession } from '../context/SessionContext';
import useRequest from '../hooks/useRequest';
import '../assets/styles/components/TestsDownload.scss';

function TestsDownload() {
  const { session } = useSession();
  const [errorArray, setErrorArray] = useState(false);
  const [loadingTests, setLoadingTests] = useState(false);
  const [tests, setTests] = useState({});
  const [errorTests, setErrorTests] = useState(null);
  const testSelected = [];
  const { response, loading, error } = useRequest(
    session.token,
    `${process.env.NEXT_PUBLIC_API_URL}/orders/?username=${session.user.username}&isComplete=true`,
    1,
  );

  function checkTest(testID) {
    if (testSelected.includes(testID)) {
      const indexDelete = testSelected.indexOf(testID);
      testSelected.splice(indexDelete, 1);
    } else {
      testSelected.push(testID);
    }
  }

  async function downloadTests() {
    if (!testSelected.length) {
      setErrorArray(true);
    } else {
      setErrorArray(false);
      setLoadingTests(true);
      setErrorTests(null);
      try {
        const URL = `${process.env.NEXT_PUBLIC_API_URL}/pdfs`;
        const object = {
          orderIds: testSelected,
        };
        const config = {
          headers: { Authorization: `Bearer ${session.token}` },
        };
        const res = await axios.post(URL, object, config);
        setTests(res);
      } catch (err) {
        setErrorTests(err);
      }
      setLoadingTests(false);
    }
  }

  if (loading) {
    return <div className="loader" />;
  }

  return (
    <>
      {response && response.data ? (
        <div className="tests-download">
          <div className="tests-download__options">
            <span className="input__error">
              {errorArray ? 'Check tests to download' : ''}
            </span>
            <button className="btn" type="button" onClick={downloadTests}>
              Download select tests
            </button>
          </div>
          <table className="table">
            <thead className="table__head">
              <tr className="table__head__row">
                <th className="table__head__row__cell">
                  <strong>Select</strong>
                </th>
                <th className="table__head__row__cell">
                  <strong>Test</strong>
                </th>
                <th className="table__head__row__cell">
                  <strong>Result Date</strong>
                </th>
              </tr>
            </thead>

            <tbody className="table__body">
              {response.data.map((test) => (
                <tr key={test._id} className="table__body__row">
                  <td className="table__body__row__cell__user">
                    {test.isComplete ? (
                      <input
                        name="check"
                        type="checkbox"
                        onChange={() => {
                          checkTest(test._id);
                        }}
                      />
                    ) : (
                      ''
                    )}
                  </td>
                  <td className="table__body__row__cell">
                    <span>{test.name}</span>
                  </td>
                  <td className="table__body__row__cell">
                    <span>{test.resultDate}</span>
                  </td>
                </tr>
              ))}
            </tbody>
          </table>
        </div>
      ) : (
        ''
      )}
    </>
  );
}

export default TestsDownload;
