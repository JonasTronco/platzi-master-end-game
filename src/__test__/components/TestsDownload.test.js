import TestsDownload from '../../components/TestsDownload';
import { create } from 'react-test-renderer';
import Router from 'next/router';
import { mount } from 'enzyme';
import '@testing-library/jest-dom';
import { SessionProvider } from '../../context/SessionContext';

describe('TestsDownload Component', () => {
  const mockedRouter = { push: () => {} };
  Router.router = mockedRouter;
  const obj = {
    session: {
      token: '',
      user: {
        id: '',
        username: '',
        typeOfUser: '',
        isActive: true,
        imageURL: '',
        firstName: '',
        lastName: '',
        defaultPath: ''
      }
    }
  };

  test('Should render TestsDownload component', () => {
    const component = create(
      <SessionProvider value={obj}>
        <TestsDownload />
      </SessionProvider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('<Tests Download />', () => {
    const component = mount(
      <SessionProvider value={obj}>
        <TestsDownload />
      </SessionProvider>
    );
    expect(component.length).toEqual(1);
  });
});
