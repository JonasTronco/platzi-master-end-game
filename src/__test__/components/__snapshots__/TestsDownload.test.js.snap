// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestsDownload Component Should render TestsDownload component 1`] = `""`;
